### PR TITLE
feat: Mass Minting in drop page

### LINKED ISSUE #9136: Mass Minting in drop page
we've been briefly chatting about this feature as it would be worth if someone want to buy from one collection right away like 10 pieces and don't want to wait long.

Would be nice to materialize this so we can speed up mints and don't need to wait 1-by-1 minting

# Design 
image

after: 
- #9171 

file: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?type=design&node-id=394-6126&mode=design&t=k6t3uvZZb3DCxEFi-11

### PR DESCRIPTION
**Pull request by hassnian**
## PR Type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

## Context

- [x] Closes #9136
- [x] Closes https://github.com/kodadot/private-workers/issues/84

## Needs QA check

- @kodadot/qa-guild please review

- amount of nfts that you can mint depends on:
    - paid: only (balance / price)
    - holderOf: min between (balance / price) and amount of available nfts 

- when minting one nft , ensure it's not generating a new one , ie it's using the same 

CleanShot 2024-03-04 at 13 26 32@2x

- when multiple nfts are being generated , no duplicates should be generated
- **paid , paid + holderOf , holdeof** , mass mint code has been added  , **free** code refactoring but still should be tested 

## Needs Design check

- @exezbcz please review

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://kodadot.xyz/dot/transfer?target=16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD)

## Screenshot 📸

- [ ] My fix has changed **something** on UI;

### CODE REVIEW

**Comment by yangwao** [comment]
Will we make it to release? That would be amazing if so

**Comment by hassnian** [comment]
that was my aim from the get-go, will try , but I doubt it 

I'd want @prury to be able to test it well since a lot of drop page code is changing.

also found out this https://github.com/kodadot/private-workers/issues/84 is not implemented and it's needed.

**Comment by yangwao** [comment]
Okay 👌

Let's aim for the next release

**Comment by yangwao** [comment]
Time to test it?

**Comment by hassnian** [comment]
Once we have https://github.com/kodadot/private-workers/pull/86 merged

**Comment by hassnian** [comment]
@prury  this is important 
CleanShot 2024-03-03 at 14 37 29@2x

otherwise each time you  massmint you'll be allocated new sn numbers

also all drop types have been received changes , free, paid , HolderOf, so once we have the worker ready please test them 

thanks

**Comment by hassnian** [comment]
if you see the same sn , like this 
CleanShot 2024-03-04 at 16 53 18@2x

it's because the worker is not deployed

**Comment by prury** [comment]
buttons keeps loading even if i did not click:
https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c

still not ready for testing i guess?
 
https://github.com/kodadot/nft-gallery/assets/36627808/44d252b2-409a-4afb-a196-396ac14dcc16

**Comment by hassnian** [comment]
> buttons keeps loading even if i did not click: https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c
> 

checking

> still not ready for testing i guess? https://github.com/kodadot/nft-gallery/assets/36627808/44d252b2-409a-4afb-a196-396ac14dcc16
> 
> second try: https://github.com/kodadot/nft-gallery/assets/36627808/6d1c807d-e785-4400-8990-a71173f269f3

yep , the worker is not deployed cc @preschian

**Comment by preschian** [comment]
@prury @hassnian I just deployed the worker, can check it again 🙏🏻

File: composables/drop/useDropMassMint.ts [hunk 1]
```diff
@@ -0,0 +1,327 @@
+import {
+  AllocateCollectionResult,
```
Comment (preschian) [inline composables/drop/useDropMassMint.ts:2]: missing?

File: components/collection/drop/GenerativeLayout.vue [hunk 2]
```diff
@@ -102,10 +106,13 @@ import type {
 import { useCollectionMinimal } from '@/components/collection/utils/useCollectionDetails'
 import useCursorDropEvents from '@/composables/party/useCursorDropEvents'
 import { DropEventType } from '@/composables/party/types'
+import { GenerativePreviewItem } from '@/composables/drop/useGenerativePreview'
 
 const props = withDefaults(
   defineProps<{
     collectionId: string
+    amountToMint?: number
```
Comment (preschian) [inline components/collection/drop/GenerativeLayout.vue:114]: eslint "requires default value to be set"

**Review by preschian — APPROVED** [review]
otherwise lgtm

File: composables/drop/useDropMassMint.ts (hunk 1 above)
Comment (hassnian) [reply composables/drop/useDropMassMint.ts:2]: yup, mb missed it , fixed

File: components/collection/drop/GenerativeLayout.vue (hunk 2 above)
Comment (hassnian) [reply components/collection/drop/GenerativeLayout.vue:114]: fixed

**Comment by hassnian** [comment]
> > buttons keeps loading even if i did not click: https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c
> 
> checking
> 

can't reproduce , it's still happening?. drop name ?

**Comment by exezbcz** [comment]
i cant set the number when i am not connected? 


https://github.com/kodadot/nft-gallery/assets/90852205/e3a0f8ff-f930-44ef-b63a-8ad7d423db02

is it feature?

**Comment by hassnian** [comment]
> i cant set the number when i am not connected?
> 
>  Screen.Recording.2024-03-05.at.12.18.48.mov 
> is it feature?
> 
> actually does not work even when connected
> 
>  Screen.Recording.2024-03-05.at.12.20.00.mov

had a chat with @exezbcz no balance checking should be done for paid drops , the user can add more funds inside the modal and be able to mint , so no max mint amount for paid drops, fixing

**Comment by hassnian** [comment]
> > > buttons keeps loading even if i did not click: https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c
> > 
> > 
> > checking
> 
> can't reproduce , still happening?. drop name ?

@prury found the issue , fixing it

**Comment by prury** [comment]
asked for everyone's review since big

**Comment by prury** [comment]
issues found: 
- preview generation still won't stop
- thumbnails not appearing on success modal
- copy button leads to only one of them(i suppose best its to lead to person profile?)
- listing does not show numbering(not sure if from this PR)
- if listing is cancelled nfts stays on blue square selection state
- thumbnails in latest nft mints took a while to be updated 5+ minutes

https://github.com/kodadot/nft-gallery/assets/36627808/80f4c6ab-f0a7-4ca4-b6c2-7fbe84d1bdbc

**Comment by daiagi** [comment]
@hassnian 
I think this PR should wait until https://github.com/kodadot/nft-gallery/issues/9575#issuecomment-1980156973
are done
then it should be re-written against https://github.com/kodadot/nft-gallery/tree/drops-refactor branch

**Comment by hassnian** [comment]
> 
> * preview generation still won't stop
> * thumbnails not appearing on success modal

fixed

> * copy button leads to only one of them(i suppose best its to lead to person profile?)

had a chat with @exezbcz regarding this when I worked on #8978 , we decided that the copy url button should be removed in the case of multiple nfts, 

fixed

> * listing does not show numbering(not sure if from this PR)

nope not from this pr,  please raise a new issue , seems  related to https://github.com/kodadot/nft-gallery/issues/9144

> * if listing is cancelled nfts stays on blue square selection state

fixed

> * thumbnails in latest nft mints took a while to be updated 5+ minutes

iirc the metadata updating process is queued, not sure cc @vikiival

**Comment by hassnian** [comment]
> @hassnian I think this PR should wait until [#9575 (comment)](https://github.com/kodadot/nft-gallery/issues/9575#issuecomment-1980156973) are done then it should be re-written against https://github.com/kodadot/nft-gallery/tree/drops-refactor branch

@daiagi  got no issues with that,  feel free to add the s-blocked tag , will come back to this pr once those issues are done

**Comment by yangwao** [comment]
Good to put in hold to prevent some bad things from happening, but it's been a month since issue was raised, let's aim please for next release or we can do post-release if everything will be good to go

**Comment by vikiival** [comment]
> iirc the metadata updating process is queued, not sure cc @vikiival

You can use batch uploading. 

So you batch upload all images and then batch upload all metadatas. 

Same is used by mass mint. 
@daiagi can help with this one

**Comment by yangwao** [comment]
I guess once this is merged
- https://github.com/kodadot/nft-gallery/pull/9620
we can push this one as well?

**Comment by vikiival** [comment]
> we can push this one as well?

Mass upload of metadata is not implemented, see the message above

**Comment by hassnian** [comment]
> > we can push this one as well?
> 
> Mass upload of metadata is not implemented, see the message above

on it

**Comment by yangwao** [comment]
- https://github.com/kodadot/nft-gallery/pull/9620
Close bein done, can be adjusted towards that branch and aim for next week release 🙏

**Comment by hassnian** [comment]
@yangwao  the target of that pr is not main 

CleanShot 2024-03-11 at 08 20 07@2x

CleanShot 2024-03-11 at 08 13 09@2x

should I target `drops-refactor` ?

**Comment by hassnian** [comment]
> @hassnian
> 
> image
> 
> PR against the refactor branch please



> @hassnian
> 
> image
> 
> PR against the refactor branch please

will be branching out from drops-refactor and adding this code manually , too many conflicts.  don't want to break something while merging

**Comment by daiagi** [comment]
close this one then?

**Pull request closed without merging**